### PR TITLE
Hw 4

### PR DESCRIPTION
**Pull request by Fedich09**

### CODE REVIEW

File: src/main/java/mate/dao/impl/CarDaoImpl.java [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+        Storage.addCar(car);
+        return car;
+    }
+
+    @Override
+    public Car get(Long id) {
+        return Storage.cars.stream()
+                .filter(e -> e.getId().equals(id))
+                .findFirst()
+                .orElseThrow(() -> new RuntimeException("Car not exist!"));
```
Comment (Ketch92) [inline src/main/java/mate/dao/impl/CarDaoImpl.java]: a grammatical error in the message "Car doesn't exist!"

File: src/main/java/mate/dao/impl/CarDaoImpl.java [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+                driver = Storage.drivers.get(i);
+            }
+        }
+        for (int i = 0; i < Storage.cars.size(); i++) {
+            for (int j = 0; j < Storage.cars.get(i).getDrivers().size(); j++) {
+                if (Storage.cars.get(i).getDrivers().get(j).equals(driver)) {
+                    cars.add(Storage.cars.get(i));
+                }
+            }
+        }
```
Comment (Ketch92) [inline src/main/java/mate/dao/impl/CarDaoImpl.java]: hmmm... I would suggest using streams here

File: src/main/java/mate/dao/impl/DriverDaoImpl.java [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+        Storage.addMDriver(driver);
+        return driver;
+    }
+
+    @Override
+    public Driver get(Long id) {
+        return Storage.drivers.stream()
+                .filter(e -> e.getId().equals(id))
+                .findFirst()
+                .orElseThrow(() -> new RuntimeException("Drive not exist!"));
```
Comment (Ketch92) [inline src/main/java/mate/dao/impl/DriverDaoImpl.java]: message "Driver doesn't exist!"

**Review by Ketch92 — CHANGES_REQUESTED** [review]

File: src/main/java/mate/dao/impl/CarDaoImpl.java (hunk 2 above)
Comment (mfedenko09) [reply src/main/java/mate/dao/impl/CarDaoImpl.java]: Good idea, I tried, but  I couldn't

**Review by Ketch92 — APPROVED** [review]

File: src/main/java/mate/service/impl/DriverServiceImpl.java [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+import mate.dao.DriverDao;
+import mate.lib.Inject;
+import mate.lib.Service;
+import mate.model.Driver;
+import mate.service.DriverService;
+
+@Service
+public class DriverServiceImpl implements DriverService {
+    @Inject
+    DriverDao driverDao;
```
Comment (RomanDubovskyi) [inline src/main/java/mate/service/impl/DriverServiceImpl.java]: access modifiers

File: src/main/java/mate/dao/impl/CarDaoImpl.java [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+@Dao
+public class CarDaoImpl implements CarDao {
+    @Override
+    public Car create(Car car) {
+        Storage.addCar(car);
+        return car;
+    }
+
+    @Override
+    public Car get(Long id) {
```
Comment (RomanDubovskyi) [inline src/main/java/mate/dao/impl/CarDaoImpl.java]: we should return optional from dao layer

File: src/main/java/mate/dao/impl/CarDaoImpl.java [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+        return car;
+    }
+
+    @Override
+    public boolean delete(Long id) {
+        return Storage.cars.removeIf(e -> e.getId().equals(id));
+    }
+
+    @Override
+    public void addDriverToCar(Driver driver, Car car) {
```
Comment (RomanDubovskyi) [inline src/main/java/mate/dao/impl/CarDaoImpl.java]: this method should be on the service layer

File: src/main/java/mate/dao/impl/CarDaoImpl.java [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+        List<Driver> cars = car.getDrivers();
+        cars.add(driver);
+        car.setDrivers(cars);
+    }
+
+    @Override
+    public void removeDriverFromCar(Driver driver, Car car) {
```
Comment (RomanDubovskyi) [inline src/main/java/mate/dao/impl/CarDaoImpl.java]: same here

File: src/main/java/mate/dao/impl/CarDaoImpl.java (hunk 2 above)
Comment (RomanDubovskyi) [reply src/main/java/mate/dao/impl/CarDaoImpl.java]: return Storage.cars.stream()
                .filter(c -> c.getDrivers().stream()
                    .anyMatch(d -> d.getId().equals(driverId)))
                .collect(Collectors.toList());

File: src/main/java/mate/dao/impl/DriverDaoImpl.java [hunk 8]
```diff
@@ -0,0 +1,44 @@
...
+@Dao
+public class DriverDaoImpl implements DriverDao {
+    @Override
+    public Driver create(Driver driver) {
+        Storage.addMDriver(driver);
+        return driver;
+    }
+
+    @Override
+    public Driver get(Long id) {
```
Comment (RomanDubovskyi) [inline src/main/java/mate/dao/impl/DriverDaoImpl.java]: same here

File: src/main/java/mate/service/ManufacturerService.java [hunk 9]
```diff
@@ -0,0 +1,18 @@
+package mate.service;
+
+import java.util.List;
+import mate.model.Manufacturer;
+
+public interface ManufacturerService {
```
Comment (RomanDubovskyi) [inline src/main/java/mate/service/ManufacturerService.java]: this file is from prev hw, it should not be in current PR

File: src/main/java/mate/service/impl/CarServiceImpl.java [hunk 10]
```diff
@@ -0,0 +1,55 @@
...
+import mate.lib.Inject;
+import mate.lib.Service;
+import mate.model.Car;
+import mate.model.Driver;
+import mate.service.CarService;
+
+@Service
+public class CarServiceImpl implements CarService {
+    @Inject
+    CarDao carDao;
```
Comment (RomanDubovskyi) [inline src/main/java/mate/service/impl/CarServiceImpl.java]: what about access modifiers

**Review by RomanDubovskyi — CHANGES_REQUESTED** [review]

File: src/main/java/mate/dao/impl/CarDaoImpl.java [hunk 11]
```diff
@@ -0,0 +1,53 @@
...
+    @Override
+    public boolean delete(Long id) {
+        return Storage.cars.removeIf(e -> e.getId().equals(id));
+    }
+
+    @Override
+    public List<Car> getAllByDriver(Long driverId) {
+        return Storage.cars.stream()
+                .filter(c -> c.getDrivers().stream()
+                        .anyMatch(d -> d.getId() != null && d.getId().equals(driverId)))
```
Comment (RomanDubovskyi) [inline src/main/java/mate/dao/impl/CarDaoImpl.java:50]: this is optional, but better to use Objects.equals

File: src/main/java/mate/service/impl/CarServiceImpl.java [hunk 12]
```diff
@@ -0,0 +1,67 @@
...
+    @Override
+    public void removeDriverFromCar(Driver driver, Car car) {
+        List<Driver> drivers = new ArrayList<>();
+        for (int i = 0; i < car.getDrivers().size(); i++) {
+            if (!car.getDrivers().get(i).equals(driver)) {
+                drivers.add(car.getDrivers().get(i));
+            }
+        }
+        car.setDrivers(drivers);
+        carDao.update(car);
```
Comment (RomanDubovskyi) [inline src/main/java/mate/service/impl/CarServiceImpl.java:60]:
```suggestion
        car.getDrivers().remove(driver);
        carDao.update(car);
```
same with the add

**Review by RomanDubovskyi — CHANGES_REQUESTED** [review]
if you merge this PR, you will loose ManufacturerService from master. We don't want to delete those files from the project, point is that these files should not be in the PR. 
You probably have created PR wrongly. Try to open PR with changes that are related only to this hw

**Pull request closed without merging**